File: netsim/src/ssfnet/net/net.cc
/**
 * \file net.cc
 * \brief Source file for the Net class.
 * \author Nathanael Van Vorst
 * \author Jason Liu
 * 
 * Copyright (c) 2011 Florida International University.
 *
 * Permission is hereby granted, free of charge, to any individual or
 * institution obtaining a copy of this software and associated
 * documentation files (the "software"), to use, copy, modify, and
 * distribute without restriction.
 *
 * The software is provided "as is", without warranty of any kind,
 * express or implied, including but not limited to the warranties of
 * merchantability, fitness for a particular purpose and
 * non-infringement.  In no event shall Florida International
 * University be liable for any claim, damages or other liability,
 * whether in an action of contract, tort or otherwise, arising from,
 * out of or in connection with the software or the use or other
 * dealings in the software.
 *
 * This software is developed and maintained by
 *
 *   Modeling and Networking Systems Research Group
 *   School of Computing and Information Sciences
 *   Florida International University
 *   Miami, Florida 33199, USA
 *
 * You can find our research at http://www.primessf.net/.
 */

#include "os/logger.h"
#include "os/community.h"
#include "net/net.h"
#include <math.h>
#include "os/partition.h"
namespace prime {
namespace ssfnet {

LOGGING_COMPONENT(Net);


Net::Net() {
}

Net::~Net() {
	ChildIterator<Net*> n = subnets();
	while(n.hasMoreElements())
		delete n.nextElement();

	ChildIterator<Host*> h = hosts();
	while(h.hasMoreElements())
		delete h.nextElement();

	ChildIterator<Link*> l = links();
	while(l.hasMoreElements())
		delete l.nextElement();

	ChildIterator<RoutingSphere*> r = rsphere();
	while(r.hasMoreElements())
		delete r.nextElement();

	ChildIterator<Monitor*> m = monitors();
	while(m.hasMoreElements())
		delete m.nextElement();

	ChildIterator<Aggregate*> a = aggregates();
	while(a.hasMoreElements())
		delete a.nextElement();


	//the traffic manager will delete the traffics
}

void Net::initStateMap() {
	BaseEntity::initStateMap();
}

void Net::init() {
	LOG_INFO("Net::init(), "<<getUName()<<endl);
	ChildIterator<Net*> n = subnets();
	while(n.hasMoreElements()) {
		//n.nextElement()->init();
		Net* hh = n.nextElement();
		LOG_INFO("Net::init() initing "<<hh->getUName()<<endl);
		hh->init();
	}

	ChildIterator<Host*> h = hosts();
	while(h.hasMoreElements()) {
		//h.nextElement()->init();
		Host* hh = h.nextElement();
		LOG_INFO("Net::init() initing "<<hh->getUName()<<endl);
		hh->init();
	}

	ChildIterator<Link*> l = links();
	while(l.hasMoreElements()) {
		//l.nextElement()->init();
		Link* hh = l.nextElement();
		LOG_INFO("Net::init() initing "<<hh->getUName()<<endl);
		hh->init();
	}

	ChildIterator<RoutingSphere*> r = rsphere();
	while(r.hasMoreElements()) {
		//r.nextElement()->init();
		RoutingSphere* hh = r.nextElement();
		LOG_INFO("Net::init() initing "<<hh->getUName()<<endl);
		hh->init();
	}
	//aggs must be inited BEFORE monitors!
	ChildIterator<Aggregate*> a = aggregates();
	while(a.hasMoreElements()) {
		//r.nextElement()->init();
		Aggregate* hh = a.nextElement();
		LOG_INFO("Net::init() initing "<<hh->getUName()<<endl);
		hh->init();
	}

	ChildIterator<Monitor*> m = monitors();
	while(m.hasMoreElements()) {
		//r.nextElement()->init();
		Monitor* hh = m.nextElement();
		LOG_INFO("Net::init() initing "<<hh->getUName()<<endl);
		hh->init();
	}
	
	//the traffic manager will init the traffics
}

void Net::wrapup()
{
	ChildIterator<Net*> n = subnets();
	while(n.hasMoreElements())
		n.nextElement()->wrapup();

	ChildIterator<Link*> l = links();
	while(l.hasMoreElements())
		l.nextElement()->wrapup();

	ChildIterator<Host*> h = hosts();
	while(h.hasMoreElements())
		h.nextElement()->wrapup();

	ChildIterator<RoutingSphere*> r = rsphere();
	while(r.hasMoreElements())
		r.nextElement()->wrapup();

	ChildIterator<Monitor*> m = monitors();
	while(m.hasMoreElements())
		m.nextElement()->wrapup();

	ChildIterator<Aggregate*> a = aggregates();
	while(a.hasMoreElements())
		a.nextElement()->wrapup();


	//the traffic manager will wrapup the traffics
}

const IPPrefix& Net::getIPPrefix() {
	return unshared.ip_prefix.read();
}

void Net::exportVizState(StateLogger* state_logger, double sampleInterval){
	ChildIterator<Aggregate*> aa = aggregates();
	while(aa.hasMoreElements()) {
		Aggregate* a = aa.nextElement();
		//std::cout<<"\tstaart export viz state for "<<a->getUName()<<", uid="<<a->getUID()<<", logger="<<((void*)state_logger)<<endl;
		a->exportVizState(state_logger, sampleInterval);
		//std::cout<<"\tdone viz state for "<<a->getUName()<<", uid="<<a->getUID()<<", logger="<<((void*)state_logger)<<endl;
	}
}

void Net::exportState(StateLogger* state_logger, double sampleInterval){
	ChildIterator<Aggregate*> aa = aggregates();
	while(aa.hasMoreElements()) {
		Aggregate* a = aa.nextElement();
		if(!a->getConfigType()->isSubtype(VizAggregate::getClassConfigType())) {
			a->exportState(state_logger, sampleInterval);
		}
	}
}


CNFContentOwnerMap* Net::getCNFContentOwnerMap() {
	return &(unshared.cnf_content_ids.read());
}

uint8_t Net::getIPPrefixLen() {
	return shared.ip_prefix_len.read();
}

Net* Net::getSuperNet() { return SSFNET_STATIC_CAST(Net*,getParent()); }

ChildIterator<Monitor*> Net::getMonitors(int* nmonitors)
{
	if(nmonitors) *nmonitors = monitorsSize();
	return monitors();
}

ChildIterator<Aggregate*> Net::getAggregates(int* naggregate)
{
	if(naggregate) *naggregate = aggregatesSize();
	return aggregates();
}

ChildIterator<Host*> Net::getHosts(int* nhosts)
{
	if(nhosts) *nhosts = hostsSize();
	return hosts();
}

ChildIterator<Net*> Net::getNets(int* nnets)
{
	if(nnets) *nnets = subnetsSize();
	return subnets();
}

ChildIterator<Link*> Net::getLinks(int* nlinks)
{
	if(nlinks) *nlinks = linksSize();
	return links();
}

IPPrefixRoute::List& Net::getExternalNetworkRoutes() {
	return shared.portal_rules.read();
}

RoutingSphere* Net::getRoutingSphere()
{
	switch(rsphereSize()) {
	case 0: return NULL;
	case 1: return rsphere().nextElement();
	default:
		LOG_ERROR("The net "<<getUName()<<" has "<<rsphereSize()<<" routing spheres!"<<endl);
	}
	return NULL;
}

UID_t Net::getControllerRid(){
	return shared.controller_rid.read();
}

SubEdgeList& Net::getSubEdgeInterfaces(){
	return shared.sub_edge_ifaces.read();
}

UID_t Net::getUID() {
	return BaseEntity::getUID();
}

UID_t Net::getMinRank(BaseEntity* anchor) {
	return BaseEntity::getMinRank(anchor);
}

UID_t Net::getMinUID() {
	return BaseEntity::getMinRank(NULL);
}

bool Net::isRoutingSphere() {
	if(getRoutingSphere()) {
		//LOG_DEBUG(getUName()<<"["<<getTypeName()<<"] --> "<<getRoutingSphere()->hasRoutes()<<endl);
		return getRoutingSphere()->hasRoutes();
	}
	return false;
}

ChildIterator<Traffic*> Net::getTraffics(int* ntraffic)
{
	if(ntraffic) *ntraffic=trafficsSize();
	return traffics();
}

RoutingSphere* Net::findOwningSphere(UID_t id) {
	if(!this->containsUID(id)) return NULL;
	ChildIterator<Net*> n = subnets();
	Net* sub=NULL;
	RoutingSphere * rv=NULL;
	while(n.hasMoreElements() && !rv) {
		sub=n.nextElement();
		if(sub->containsUID(id)) {
			if(sub->isRoutingSphere()) {
				return sub->getRoutingSphere();
			}
			rv = sub->findOwningSphere(id);
		}
	}
	if(!rv && this->isRoutingSphere()) rv=this->getRoutingSphere();
	return rv;
}


} // namespace ssfnet
} // namespace prime

File: netsim/src/ssfnet/net/net.m
/**
 * \file net.h
 * \brief Header file for the Net class.
 * \author Nathanael Van Vorst
 * \author Jason Liu
 * 
 * Copyright (c) 2011 Florida International University.
 *
 * Permission is hereby granted, free of charge, to any individual or
 * institution obtaining a copy of this software and associated
 * documentation files (the "software"), to use, copy, modify, and
 * distribute without restriction.
 *
 * The software is provided "as is", without warranty of any kind,
 * express or implied, including but not limited to the warranties of
 * merchantability, fitness for a particular purpose and
 * non-infringement.  In no event shall Florida International
 * University be liable for any claim, damages or other liability,
 * whether in an action of contract, tort or otherwise, arising from,
 * out of or in connection with the software or the use or other
 * dealings in the software.
 *
 * This software is developed and maintained by
 *
 *   Modeling and Networking Systems Research Group
 *   School of Computing and Information Sciences
 *   Florida International University
 *   Miami, Florida 33199, USA
 *
 * You can find our research at http://www.primessf.net/.
 */

#ifndef __NET_H__
#define __NET_H__

#include "os/ssfnet_types.h"
#include "os/config_entity.h"
#include "os/routing.h"
#include "proto/cnf/cnf_session.h"
#include "os/traffic.h"
#include "net/host.h"
#include "net/link.h"
#include "os/placeholder.h"
namespace prime {
namespace ssfnet {

/**
 * \brief A network.
 *
 * The Net class is the container for hosts (routers), links, and
 * other sub-networks. Also, traffic and routing strategy can be
 * specified for a network.
 */
class Net: public ConfigurableEntity<Net, BaseEntity> {
public:
	friend class Partition;

	state_configuration {
		shared configurable int8_t ip_prefix_len {
			type= INT;
			default_value= "-1";
			doc_string= "The n in W.X.Y.Z/n";
		};
		configurable IPPrefix ip_prefix {
			type= OBJECT;
			default_value= "0.0.0.0/0";
			doc_string= "ip prefix of this network";
		};
		configurable CNFContentOwnerMap cnf_content_ids {
			type= OBJECT;
			default_value= "[]";
			doc_string= "a list of content id and uid. i.e. [1,100,2,200] --> [ cid:1, uid:100],[ cid:2,uid:200]";
		};

		shared configurable IPPrefixRoute::List portal_rules {
			type= OBJECT;
			default_value= "[]";
			doc_string= "a list of rules in the form [ip_prefix->(router_uid,iface_uid,iface_ip)]. They will automatically be compiled into the topnet.";
		};

		shared configurable UID_t controller_rid {
			type=INT;
			default_value="0";
			doc_string="The RID of the cnf controller.";
		};

		shared configurable SubEdgeList sub_edge_ifaces {
			type=OBJECT;
			default_value="[]";
			doc_string="A comma separated list of its child sphere's edge iface and owning_host rank pairs";
		};

		child_type<Net> subnets {
			min_count = 0;
			max_count = 0;
			doc_string = "sub-networks defined within this network";
		};
		child_type<Host> hosts {
			min_count = 0;
			max_count = 0;
			doc_string = "hosts defined within this network";
		};
		child_type<Link> links {
			min_count = 0;
			max_count = 0;
			doc_string = "links defined within this network";
		};
		child_type<PlaceHolder> placeholders{
			min_count = 0;
			max_count = 0;
			doc_string = "a node to hold the place of children which exist on remote partitions";
		};
		child_type<RoutingSphere> rsphere {
			min_count = 1;
			max_count = 1;
			doc_string = "routing sphere defined within this network";
		};
		child_type<Monitor> monitors {
			min_count = 0;
			max_count = 0;
			doc_string = "data monitors";
		};
		child_type<Aggregate> aggregates {
			min_count = 0;
			max_count = 0;
			doc_string = "aggregate statistics";
		};
		child_type<Traffic> traffics {
			min_count = 0;
			max_count = 1;
			doc_string = "traffic defined within this network";
		};
	};

public:
	/** The constructor. */
	Net();

	/** The destructor. */
	virtual ~Net();

	/**
	 * The init method is used to initialize the network once it has
	 * been created and configured by model builder. The init method
	 * will initialize all sub-networks, hosts, and links therein.
	 */
	virtual void init();

	/**
	 * The wrapup method is called at the end of the simulation. It is
	 * used to collect statistical information of the simulation run.
	 * The wrapup method will call the wrapup method of all
	 * sub-networks, hosts, and links of this network.
	 */
	virtual void wrapup();

	/**
	 * Gets the uid of the net.
	 */
	UID_t getUID();

	/**
	 * Gets the rank of the left most child child in this subtree in reference to anchor
	 */
	virtual UID_t getMinRank(BaseEntity* anchor);

	/**
	 * XXX initStateMap()
	 */
	virtual void initStateMap();

	virtual void exportVizState(StateLogger* state_logger, double sampleInterval);
	virtual void exportState(StateLogger* state_logger, double sampleInterval);

	/**
	 * Gets the uid of the left most child child in this subtree
	 */
	UID_t getMinUID();

	/** Return the IP address length. (the n in W.X.Y.Z/n) */
	uint8_t getIPPrefixLen();

	/** Get the IP prefix of this network. */
	const IPPrefix& getIPPrefix();

	/** Get the CNF content owner map of this network. */
	CNFContentOwnerMap* getCNFContentOwnerMap();

	/** Return the parent network, or NULL if this is top-level
	 network. */
	Net* getSuperNet();

	/** Return all subnets defined in this network level. */
	ChildIterator<Net*> getNets(int* nnets = 0);

	/** Return all hosts defined in this network level.*/
	ChildIterator<Host*> getHosts(int* nhosts = 0);

	/** Return all links defined in this network level. */
	ChildIterator<Link*> getLinks(int* nlinks = 0);

	/** Return the routing sphere for this network if defined. */
	RoutingSphere* getRoutingSphere();

	/** Return all monitors defined in this network level.*/
	ChildIterator<Monitor*> getMonitors(int* nmonitors);

	/** Return all aggregates defined in this network level.*/
	ChildIterator<Aggregate*> getAggregates(int* naggregate);

	/**
	* Get the controller rid of this network.
	*/
	UID_t getControllerRid();

	/**
	 * Get the pairs of edge interface rank and owning host rank of sub routing spheres
	 */
	SubEdgeList& getSubEdgeInterfaces();

	/** Return all traffics defined in this network level. */
	ChildIterator<Traffic*> getTraffics(int* ntraffic = 0);

	/**
	 *  returns true if this network contains a routingsphere
	 */
	virtual bool isRoutingSphere();

	/**
	 * \breif find the sphere which owns the node with UID id. If the node is not owned by
	 * a sub-sphere, or no sub-spheres exists, this network routing sphere is considered the
	 * owning sphere. If this network is not a routing sphere, NULL is returned..
	 */
	RoutingSphere* findOwningSphere(UID_t id);

	IPPrefixRoute::List& getExternalNetworkRoutes();

};

} // namespace ssfnet
} // namespace prime

#endif /*__NET_H__*/
